File: DAA/merge_sort.c
#include <stdio.h>
void merge(int arr[], int l, int r, int m)
{
    int i, j, k;
    int n1 = m - l + 1;
    int n2 = r - m;

    int L[n1], R[n2];
    for(int i=0; i<n1; i++)
    L[i] = arr[l + i];
    for(int j=0; j<n2; j++)
    R[j] = arr[m + j + 1];

    i = 0, j = 0, k = l;
    while(i < n1 && j < n2)
    {
        if(L[i] < R[j])
        arr[k++] = L[i++];
        else
        arr[k++] = R[j++];
    }

    while(i < n1)
    arr[k++] = L[i++];

    while(j < n2)
    arr[k++] = R[j++];
}

void merge_sort(int arr[], int l, int r)
{
    if(l < r)
    {
        int mid = l + (r - l) / 2;

        merge_sort(arr, l, mid);
        merge_sort(arr, mid + 1, r);

        merge(arr, l, r, mid);
    }
}

int main()
{
    int n;
    printf("Enter size of array:");
    scanf("%d", &n);

    int arr[n];
    for(int i=0; i<n; i++)
    {
        printf("Enter element:");
        scanf("%d", &arr[i]);
    }

    merge_sort(arr, 0, n - 1);

    printf("Sorted Elements of the array:");
    for(int i=0; i<n; i++)
    printf("%d ", arr[i]);
    printf("\n");
    
    return 0;
}

File: DAA/Merge_sort.java
import java.util.Scanner;
class Merge_sort
{
    void merge(int arr[], int l, int m, int h)
    {
        int n1 = m - l + 1;
        int n2 = h - m;
        int L[] = new int[n1];
        int R[] = new int[n2];

        for(int i=0; i<n1; i++)
        L[i] = arr[l + i];
        for(int i=0; i<n2; i++)
        R[i] = arr[m + i + 1];

        int i = 0, j = 0, k = l;
        while(i < n1 && j < n2)
        {
            if(L[i] < R[j])
            arr[k] = L[i++];

            else
            arr[k] = R[j++];

            k++;
        }

        while(i < n1)
        arr[k++] = L[i++];

        while(j < n2)
        arr[k++] = R[j++];
    }

    void merge_sort(int arr[], int l, int h)
    {
        if(l < h)
        {
            int mid = (l + h) / 2;
            merge_sort(arr, l, mid);
            merge_sort(arr, mid+1, h);
            merge(arr, l, mid, h);
        }
    }

    public static void main(String arge[])
    {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of array:");
        int n = sc.nextInt();

        int arr[] = new int[n];
        for(int i=0; i<n; i++)
        {
            System.out.print("Enter element:");
            arr[i] = sc.nextInt();
        }

        Merge_sort obj = new Merge_sort();
        obj.merge_sort(arr, 0, n-1);

        System.out.println("Sorted Array:");
        for(int i=0; i<n; i++)
        System.out.print(arr[i] + " ");

        System.out.println();
    }
}